server/controllers/convo.js

```diff
@@ -0,0 +1,172 @@
+          res.send(populated);
+        }
+      });
+  }),
+  getOneConvo: asyncHandler(async (req, res) => {
+    let user1 = mongoose.Types.ObjectId("60efa11241aa990dec93324e");
+    let user2 = mongoose.Types.ObjectId("60efa13a41aa990dec93324f");
+    Message.aggregate([{
+          $lookup: {
+            from: 'user',
+            localField: 'to',
+            foreignField: '_id',
+            as: 'toObj',
+          },
+        },
+        {
+          $lookup: {
+            from: 'user',
+            localField: 'from',
+            foreignField: '_id',
+            as: 'fromObj',
+          },
+        },
+      ])
+      .match({
+        $or: [{
+            $and: [{
+              to: user1
+            }, {
+              from: user2
+            }]
+          },
+          {
+            $and: [{
+              to: user2
+            }, {
+              from: user1
+            }]
+          },
+        ],
+      })
+      .project({
+        'toObj': 0,
+        'fromObj': 0,
+        "__v": 0
+      })
+      .exec((err, messages) => {
+        if (err) {
+          console.log(err);
+          res.setHeader('Content-Type', 'application/json');
+          res.end(JSON.stringify({
+            message: 'Failure'
+          }));
+          res.sendStatus(500);
+        } else {
+          res.send(messages);
+        }
+      });
+  }),
+  createMessage: asyncHandler(async (req, res) => {
+    let from = mongoose.Types.ObjectId(req.body.from);
+    let to = mongoose.Types.ObjectId(req.body.to);
+
+    Conversation.findOneAndUpdate({
+        recipients: {
+          $all: [{
+              $elemMatch: {
+                $eq: from
+              }
+            },
+            {
+              $elemMatch: {
+                $eq: to
+              }
+            },
+          ],
+        },
+      }, {
+        recipients: [req.body.from, req.body.to],
+        lastMessage: req.body.body,
+        date: Date.now(),
+      }, {
+        upsert: true,
+        new: true,
+        setDefaultsOnInsert: true
+      },
+      async function (err, conversation) {
+        if (err) {
+          console.log(err);
+          res.setHeader('Content-Type', 'application/json');
+          res.end(JSON.stringify({
+            message: 'Failure'
+          }));
+          res.sendStatus(500);
+        } else {
+          if (!conversation._id) {
+            let newConversation = new Conversation({recipients: [req.body.from, req.body.to], lastMessage: req.body.body})
+            const { _id } = await newConversation.save()
+            conversation._id = _id;
+          }
+          let message = new Message({
+            conversation: conversation._id,
+            to: req.body.to,
+            from: req.body.from,
+            body: req.body.body,
+          });
+
+          // req.io.sockets.emit('messages', req.body.body);
```

Comment:
Please remove

Reply:
This was an intentional comment so Brian can easily find where to integrate his sockets. Would you mind if I leave it in?

Reply:
Sure, it's okay to leave it in.

Reply:
Excellent thank you